### PR TITLE
mgmt: hawkbit: use SMF

### LINKED ISSUE #71576: mgmt: hawkbit: rewrite hawkBit subsystem to use SMF.

### LINKED ISSUE #71691: mgmt: hawkbit: always approves the cancelation of an action
**Describe the bug**

Our hawkBit subsystem currently always approves the cancellation of a action (doing a firmware update) even if the firmware update had been completed.

**To Reproduce**
Create and assign a update via the hawkBit update server and cancel the update at the server between installing and starting the hawkbit subsystem after rebooting

**Expected behavior**
The hawkbit subsystem should check the saved action id and if it is identical with the one of the cancellation it should  deny the cancellation.

**Impact**
annoyance, will only happen in rare cases.

**Logs and console output**
NA

**Additional context**
NA

### LINKED ISSUE #71759: mgmt: hawkbit: support other tenants than default
**Is your enhancement proposal related to a problem? Please describe.**
Currently the hawkBit subsystem only supports `default` as tenant. hawkBit should also support other tenants.

**Describe the solution you'd like**
Be able to set it via a Kconfig option.

**Describe alternatives you've considered**
NA

**Additional context**
Most SaaS Products, that include hawkBit for updating (like https://bosch-iot-suite.com/service/rollouts/) require changing the tenant, as they don't use the `default` tenant

### PR DESCRIPTION
**Pull request by maass-hamburg**
Use the [State Machine Framework](https://docs.zephyrproject.org/latest/services/smf/index.html) for hawkBit. Includes a lot of cleanup of the hawkBit subsystem.

Resolves: #71576 
Resolves: #71691
Resolves: #71759

Depends on/the following PRs should be merged before this: 
- https://github.com/zephyrproject-rtos/zephyr/pull/68806
- https://github.com/zephyrproject-rtos/zephyr/pull/69293
- https://github.com/zephyrproject-rtos/zephyr/pull/70386
- https://github.com/zephyrproject-rtos/zephyr/pull/71037
- #71250
- #72483

### CODE REVIEW

**Comment by kartben** [comment]
cc @glenn-andrews

File: subsys/mgmt/hawkbit/hawkbit.c
Comment (glenn-andrews) [inline subsys/mgmt/hawkbit/hawkbit.c:932]: I have a dislike of this construct as it doesn't follow UML rules for state machines, and I'm hoping to put in an RFC for making SMF UML-compliant.

Would it be possible to make an error-handling state outside the other states and at the same level as the root state, that you transition to on an error? This would cause all exit actions to be called, plus the entry action of the error-handler state, which can set_smf_terminate() in its entry action.

**Review by glenn-andrews — CHANGES_REQUESTED** [review]
Consider transitioning to an error-handling state rather than using `smf_set_state(SMF_CTX(s), NULL);`

File: subsys/mgmt/hawkbit/hawkbit.c
Comment (glenn-andrews) [reply subsys/mgmt/hawkbit/hawkbit.c]: I think so, I'm looking on my phone so not able to do a full examination.

Is s_terminate the entry action or the run action? If it's the run action you'd need to post an event to get it to run, I think?

Comment (maass-hamburg) [reply subsys/mgmt/hawkbit/hawkbit.c]: it's an entry action.

Comment (glenn-andrews) [reply subsys/mgmt/hawkbit/hawkbit.c]: Sounds good. Thank you.

**Comment by real-tintin** [comment]
This is great! We have considered rewriting the implementation to use a state machine due to multiple issue (I previously wrote in the roadmap issue). As I wrote there, we have made adaptions and bug fixes e.g., issues with double freeing and handling of NULL response (crashes hard during download if not checked). But I see that you have improved on the malloc and I strongly believe that the SMF will solve a bunch of issues itself.

I would like to give this a try. I will merge (locally for testing) our adaptions e.g., a dynamic config e.g., setting endpoints and tokens (during a device provisioning over BLE) - I saw that it already exist a PR for a similar change.

**Comment by maass-hamburg** [comment]
> This is great! We have considered rewriting the implementation to use a state machine due to multiple issue (I previously wrote in the roadmap issue). As I wrote there, we have made adaptions and bug fixes e.g., issues with double freeing and handling of NULL response (crashes hard during download if not checked). But I see that you have improved on the malloc and I strongly believe that the SMF will solve a bunch of issues itself.
> 
> I would like to give this a try. I will merge (locally for testing) our adaptions e.g., a dynamic config e.g., setting endpoints and tokens (during a device provisioning over BLE) - I saw that it already exist a PR for a similar change.

If you find bugs, please feel free to open an issue. And if you have a fix for it, it would be wonderful if you [contribute it back](https://docs.zephyrproject.org/latest/contribute/index.html) into the zephyr main. It makes thinks easier for all of us. 

The same is true for new features, feel free to contribute. Just look at the features we added to hawkBit in the last weeks: [here](https://github.com/zephyrproject-rtos/zephyr/pulls?q=hawkbit)

File: subsys/mgmt/hawkbit/hawkbit.c
Comment (glenn-andrews) [inline subsys/mgmt/hawkbit/hawkbit.c:1489]: Can this ever be zero? e.g. with a successful download?

Comment (maass-hamburg) [reply subsys/mgmt/hawkbit/hawkbit.c:1428]: No

Comment (glenn-andrews) [reply subsys/mgmt/hawkbit/hawkbit.c:1427]: > Yes it does, but not including today's changes.

Heh. okay.

**Comment by real-tintin** [comment]
@maass-hamburg while you are at it. Could you also remove the hard coded CA_CERTIFICATE_TAG (https://github.com/zephyrproject-rtos/zephyr/blob/c13061f72e7d2f6dae7e2e5cfd3e904aa303f8e0/subsys/mgmt/hawkbit/hawkbit.c#L37) and instead use the dynamic config? 

We have done this as we have multiple certs/tags which are configured elsewhere. Maybe this could be config option if someone wants to keep hard coded value (1) e.g., CONFIG_HAWKBIT_USE_STATIC_CERT_TAG (with a CONFIG_HAWKBIT_STATIC_CERT_TAG) vs. CONFIG_HAWKBIT_USE_DYNAMIC_CERT_TAG.

**Comment by maass-hamburg** [comment]
> @maass-hamburg while you are at it. Could you also remove the hard coded CA_CERTIFICATE_TAG:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/c13061f72e7d2f6dae7e2e5cfd3e904aa303f8e0/subsys/mgmt/hawkbit/hawkbit.c#L37
> 
> and instead use the dynamic config?
> We have done this as we have multiple certs/tags which are configured elsewhere. Maybe this could be config option if someone wants to keep hard coded value (1) e.g., CONFIG_HAWKBIT_USE_STATIC_CERT_TAG (with a CONFIG_HAWKBIT_STATIC_CERT_TAG) vs. CONFIG_HAWKBIT_USE_DYNAMIC_CERT_TAG.

@real-tintin https://github.com/zephyrproject-rtos/zephyr/pull/70787/commits/78ab1ed0eadf6ba27530a1011d20131855cb0573 please test if it works. I might put it in a separate PR, when I put this PR out of WIP, so this PR dosen't get to big , as that feature is not dependant on the smf and not a priority for me.

**Comment by maass-hamburg** [comment]
> @maass-hamburg I have integrated your changes and will start testing with our Hawkbit setup tomorrow. I added a smaller suggestion.
> 
> I also gathered some other points (which are not caused by your changes):
> 
> 1. Add callback register for completed probe with the resulting status code e.g., store diagnostic error counters or reboot the device through a common interface (we do a cold reboot) e.g., in header
> 
> ```
> /**
>  * @typedef hawkbit_probe_cb_t
>  * @brief Callback for handling hawkBit probe events.
>  *
>  * Is called for a probe with the resulting response.
>  *
>  * @param response The hawkBit response.
>  */
> typedef void (*hawkbit_probe_cb_t)(const enum hawkbit_response response);
> 
> /**
>  * @brief Register a callback for hawkBit probe events.
>  *
>  * @note Only one cb is supported and the registered callback
>  * shouldn't block for too long as it will effectively block
>  * the scheduling of future probing.
>  *
>  * @param cb A pointer to the callback to register.
>  */
> void hawkbit_register_probe_cb(hawkbit_probe_cb_t cb);
> ```
> 
> and in the source
> 
> ```
> static void autohandler(struct k_work *work)
>     ....
>     if (hawkbit_probe_cb != NULL) {
>         hawkbit_probe_cb(response);
>     }
> }
> 
> void hawkbit_register_probe_cb(hawkbit_probe_cb_t cb)
> {
>     hawkbit_probe_cb = cb;
> }
> ```
> 
> This would also need some config to disable the warm reboot done during the probe respone checking e.g., HAWKBIT_HANDLE_REBOOT.
> 
> 2. Rename DEVICE_ID_HEX_MAX_SIZE to just DEVICE_ID_MAX_LEN (not just MAC-address).
> 3. Add doxygen return type to hawkbit_get_device_identity_cb_handler_t.
> 
> If you feel up for it, feel free to add whatever you want of these. Otherwise I can open a PR afterwards. The first one is obviously the most important one.

that would be a good things be be part of your own PRs

File: subsys/mgmt/hawkbit/hawkbit.c
Comment (real-tintin) [inline subsys/mgmt/hawkbit/hawkbit.c:1187]: It seems like the resulting url is missing the "actionId" and the ending "feedback".

I believe that "actionId" is unintentionally removed by hawkbit_find_cancel_action_id which uses strtok, which in turn is modifying the str in place. A solution would be to copy href to a buffer before modifying it.

File: subsys/mgmt/hawkbit/hawkbit.c
Comment (real-tintin) [inline subsys/mgmt/hawkbit/hawkbit.c:1310]: Suggestion: "Ready to download update"

**Comment by real-tintin** [comment]
Tested the recent changes, works fine! I guess I don't have the buttons to approve this... 

Awesome work, feel much more rigid now! 🙂

**Comment by real-tintin** [comment]
Integrated latest changes and tested them, this includes:

- No update
- Cancel of update
- Download and image swap

No issues.

We have a setup where devices are OTA updated every X h, stress test. I will create a alpha release with your latest changes and run in the setup. I can get back to you with the result after the weekend.

**Comment by maass-hamburg** [comment]
> Integrated latest changes and tested them, this includes:
> 
> * No update
> * Cancel of update
> * Download and image swap
> 
> No issues.
> 
> We have a setup where devices are OTA updated every X h, stress test. I will create a alpha release with your latest changes and run in the setup. I can get back to you with the result after the weekend.

thats sounds good

**Comment by real-tintin** [comment]
The devices have OTA updated as expected over the weekend. I saw that you pushed some changes today. Should these affect the SMF?

**Comment by maass-hamburg** [comment]
> The devices have OTA updated as expected over the weekend. I saw that you pushed some changes today. Should these affect the SMF?

It shouldn't. See them here: https://github.com/zephyrproject-rtos/zephyr/compare/cd2578014f1965de9747b82ac29cc929504ee273..93ba8d566c509b1b8b6dd383d751843ae10386f2

**Review by real-tintin — APPROVED** [review]
Looks good to me. I have performed integration testing - multiple devices getting OTA updated every 1 h, over one weekend.

File: doc/releases/release-notes-3.7.rst
Comment (maass-hamburg) [reply doc/releases/release-notes-3.7.rst:304]: Yes, sheduling the autohandler, that only run once, will not interfere with the other autohandler. btw these changes are from https://github.com/zephyrproject-rtos/zephyr/pull/71037

File: doc/releases/release-notes-3.7.rst
Comment (maass-hamburg) [reply doc/releases/release-notes-3.7.rst:305]: btw these changes are from https://github.com/zephyrproject-rtos/zephyr/pull/71037

File: doc/releases/release-notes-3.7.rst
Comment (maass-hamburg) [reply doc/releases/release-notes-3.7.rst:308]: btw these changes are from https://github.com/zephyrproject-rtos/zephyr/pull/71037

File: doc/releases/release-notes-3.7.rst
Comment (maass-hamburg) [reply doc/releases/release-notes-3.7.rst:312]: btw these changes are from https://github.com/zephyrproject-rtos/zephyr/pull/71037

File: include/zephyr/mgmt/hawkbit.h
Comment (real-tintin) [inline include/zephyr/mgmt/hawkbit.h:48]: It would be good if we could identify the type of error event. In our case we want to know if it was a connection/network issue.

File: include/zephyr/mgmt/hawkbit.h
Comment (real-tintin) [inline include/zephyr/mgmt/hawkbit.h:50]: I think that the naming of these _INSTALLED enums (response as well) are somewhat misleading, as we have two reboot and swap images before we can consider an update to be installed.

Suggestion: Rename to _DOWNLOADED.

**Comment by real-tintin** [comment]
Hi! What is the status of this? Is it possible to get this into the 3.7 release? FYI, we have already rolled this out in our prod and done rigorous real-world-testing prior to it.

**Comment by maass-hamburg** [comment]
> Hi! What is the status of this? Is it possible to get this into the 3.7 release? FYI, we have already rolled this out in our prod and done rigorous real-world-testing prior to it.

If #71037 gets approved and merged, we could start with this. 

@ycsin what do you think?

**Comment by real-tintin** [comment]
@glenn-andrews what is your opinion on this? I would very much see that this is merged. Huge improvement to the current implementation.

**Comment by glenn-andrews** [comment]
> @glenn-andrews what is your opinion on this? I would very much see that this is merged. Huge improvement to the current implementation.

Love to see it merged, but that's an @maass-hamburg question. I can comment on the HSM implementation, but someone else needs to confirm operation since I don't have any hardware to test it on.

**Comment by maass-hamburg** [comment]
@real-tintin @glenn-andrews I would love to get this merged. But https://github.com/zephyrproject-rtos/zephyr/pull/71037 needs to be merged before. Unfortunately I am now the only one responsable for the hawkBit subsystem. To get the PR moved forward, I need two approving reviews from people that are maintainer or collaborator in the mintainers.yml. If you are thinking about becoming a colaborater, go forward, but you should have contributed to the hawkbit subsystem before and maybe some other contributions in other parts of zephyr.

**Comment by glenn-andrews** [comment]
> @real-tintin @glenn-andrews I would love to get this merged. But #71037 needs to be merged before. Unfortunately I am now the only one responsable for the hawkBit subsystem. To get the PR moved forward, I need two approving reviews from people that are maintainer or collaborator in the mintainers.yml.

You may want to advertise it in the Discord server `pr-help` channel?

**Comment by maass-hamburg** [comment]
> > @real-tintin @glenn-andrews I would love to get this merged. But #71037 needs to be merged before. Unfortunately I am now the only one responsable for the hawkBit subsystem. To get the PR moved forward, I need two approving reviews from people that are maintainer or collaborator in the mintainers.yml.
> 
> You may want to advertise it in the Discord server `pr-help` channel?

did it already more than once.

**Comment by real-tintin** [comment]
> > @glenn-andrews what is your opinion on this? I would very much see that this is merged. Huge improvement to the current implementation.
> 
> Love to see it merged, but that's an @maass-hamburg question. I can comment on the HSM implementation, but someone else needs to confirm operation since I don't have any hardware to test it on.

I have tested it on real HW. As a fact, we already run it in production on +200 devices. No issues and prior to this I performed various manual testing e.g., multiple updates, no internet, power cycle during downloading etc.

I feel like no one is really responsible for the Hawkbit module? I would vote for @maass-hamburg (I”m glad to be 2nd in line, done a few smaller contributions so far).

**Review by glenn-andrews — DISMISSED** [review]
LGTM from an HSM standpoint and @real-tintin has tested it on production hardware.

I would suggest rebasing on the latest mainline and ensuring all tests work.

**Comment by maass-hamburg** [comment]
@real-tintin I did a rebase, you might want to test it again

**Comment by real-tintin** [comment]
> @real-tintin I did a rebase, you might want to test it again

Sure! I can do that next week. I don't have access to HW atm.

File: include/zephyr/mgmt/hawkbit/hawkbit.h
Comment (real-tintin) [inline include/zephyr/mgmt/hawkbit/hawkbit.h:37]: Change to _DOWNLOADED, as event enum "HAWKBIT_EVENT_UPDATE_DOWNLOADED". Update comment.

**Comment by real-tintin** [comment]
I've have now done some testing. Looks good so far e.g., "image download", "aborting and resuming (via reboot or no internet)", "already updated", "polling" etc.

Note, I have made some smaller adaptions such as:
- Disable Hawkbit control of device reboot.
- Remove board name prefix from device name.
- Add method to probe for credentials.

I plan to merge these later on (after this). I guess this will have to wait for the 4.0 release.

**Comment by real-tintin** [comment]
@glenn-andrews can we merge this now, as 4.0.0 i released?

**Comment by glenn-andrews** [comment]
> @glenn-andrews can we merge this now, as 4.0.0 i released?

My assumption is yes, provided it's been tested under 4.0.0 and we rerun CI tests.

**Comment by kartben** [comment]
you'll need to please rebase on latest main to fix CI issue

**Comment by maass-hamburg** [comment]
@glenn-andrews is rebased and ci was also successful

File: subsys/mgmt/hawkbit/hawkbit.c
Comment (nordicjm) [inline subsys/mgmt/hawkbit/hawkbit.c]: declare variables together at top of function

File: subsys/mgmt/hawkbit/hawkbit.c
Comment (nordicjm) [inline subsys/mgmt/hawkbit/hawkbit.c:1230]: could all be `else if`

File: subsys/mgmt/hawkbit/hawkbit.c
Comment (nordicjm) [inline subsys/mgmt/hawkbit/hawkbit.c]: as above

File: subsys/mgmt/hawkbit/hawkbit.c
Comment (nordicjm) [inline subsys/mgmt/hawkbit/hawkbit.c:1484]: and for all instances

File: include/zephyr/mgmt/hawkbit/event.h
Comment (nordicjm) [inline include/zephyr/mgmt/hawkbit/event.h:93]: zephyr line length is 100 characters, this can all fit on one line

**Review by nordicjm — DISMISSED** [review]
Should be tidied up as per comments but looks ok

**Comment by maass-hamburg** [comment]
> Should be tidied up as per comments but looks ok

@nordicjm done

**Review by nordicjm — APPROVED** [review]

**Review by kartben — APPROVED** [review]

**Review by kartben — APPROVED** [review]

**Pull request merged**
